### PR TITLE
Fix: Make failing code paths diagnosable

### PR DESCRIPTION
**Pull request by ModeSevenIndustrialSolutions**
## Summary

PR 2 of the stacked series clearing the `aislop` audit findings. This one makes
failing code paths diagnosable, clearing all 23 `ai-slop/silent-recovery`
findings and the single `ai-slop/python-broad-except` finding.

Verified finding count: **144 → 119**.

> **Stacked on #645.** Base is `fix/aislop-comment-hygiene`, so this diff shows
> only this PR's changes. Merge #645 first.

### 1. `Fix: Use sys.exit instead of the builtin exit`

`exit` is injected into builtins by the `site` module as an interactive
convenience. It is absent under `python -S` and in embedded/frozen
interpreters, where these calls would raise `NameError` instead of terminating —
and every one of them is on an error path. Replaces the bare calls in
`cli/github_cli.py`, `lfidapi.py`, `openstack/cmd.py` and `shell/dco.py`.

### 2. `Fix: Include the caught error in failure logs`

Nineteen handlers logged a fixed string and discarded the exception they had
just caught. `Error reading settings file "x"` does not say whether the file was
missing, unreadable, or a directory; `in _request_post. ConnectionError` said
less than the exception it replaced.

Each handler now binds the exception and interpolates it into the log call.

Two Gerrit messages change severity to match their behaviour: a failed
sanity-check query and an unparsable API response both abort the command, so
they log at `ERROR` rather than `INFO`.

The Gerrit endpoint's `exit()` → `sys.exit()` migration lands in this commit
rather than the previous one because those calls sit inside the same handlers
being rewritten and cannot be split cleanly.

### 3. `Fix: Create directories with exist_ok`

Two sites created a directory in a `try` and caught `FileExistsError` only to log
that it already existed — which is what `exist_ok` provides. Worse, a genuine
permission or read-only-filesystem error took the same silent debug path as the
benign case. The hook directory now reports and re-raises `OSError`.

### 4. `Refactor: Import AppState unconditionally`

The `AppState` re-export sat behind `except Exception: pass`, annotated as a
circular-import guard. **No cycle exists** — `cli/state.py` imports only from the
standard library. The handler suppressed nothing it was meant to, while hiding
genuine failures and leaving a name promised by `__all__` silently absent.

### 5. `Fix: Fail when the stack cost cannot be derived` ⚠️ behaviour change

**This one deserves reviewer attention.**

`openstack stack cost` wrapped the whole calculation in `except Exception`,
logged, then printed `total: 0.0`. That is indistinguishable from a genuinely
idle stack, so a pipeline parsing the figure would silently under-report.

The per-server pricing lookup *already* degrades gracefully on its own — it
catches timeouts and URL errors from the vexxhost pricing API, warns, and
contributes `0.0` for that server. So reaching the outer handler means stack or
resource enumeration itself failed and **no** total can be derived.

It now logs and exits non-zero rather than printing a fabricated total. The
genuine empty case (a stack with no servers) still prints `total: 0.0` and exits
successfully.

### 6. `Fix: Address Copilot review feedback`

Three findings from the automated review, all valid:

- **A genuine infinite loop.** In the Docker Hub tag fetch, a malformed response
  left `still_more` and `docker_tag_url` unchanged inside `while still_more:`,
  so the same page was fetched forever, re-adding its tags each pass. It span
  silently before because the handler logged at `debug`; raising the level to
  `warning` in this branch is what exposed it. Now raises.
- `sys.exit(resp.status_code)` → `sys.exit(1)`. An unparsable body can arrive
  with any status, and an HTTP code is not a meaningful process exit code.
- Added `tests/test_openstack_stack.py`, covering the behaviour change in
  commit 5 — which had no test, exactly the gap Principle I exists to close.

## Verification

- Commits were split by hunk, and replaying the groups in sequence was
  verified to reproduce the combined tree byte-for-byte.
- `uv run pytest tests/` — 826 passed
- `prek run --all-files` — clean
- `ruff check` / `ruff format` / `mypy` / `basedpyright` — clean

## Follow-ups noted, deliberately not in scope

- `shell/dco.py` accumulates `git log` output across branches but pops only one
  trailing blank entry, so blanks from all but the last branch survive. This is
  pre-existing and unrelated to slop; flagging it rather than fixing it here.

*(An earlier version of this description claimed 19 bare `exit()` calls remained
after this PR. That was wrong — `main` had exactly 19, across five files, and
all of them are converted here. A scan of the branch finds none left.)*

### CODE REVIEW

File: lftools_uv/nexus/release_docker_hub.py [hunk 1]
```diff
@@ -354,8 +354,8 @@ def __init__(self, org_name: str, repo_name: str, repo_from_file: bool) -> None:
                         still_more = True
                     else:
                         still_more = False
-                except Exception:
-                    log.debug(f"Issue fetching tags for {combined_repo_name}")
+                except Exception as exc:
+                    log.warning("Issue fetching tags for %s: %s", combined_repo_name, exc)
```
[the remark this answers is not in the document]
Comment (ModeSevenIndustrialSolutions) [reply lftools_uv/nexus/release_docker_hub.py]: Good catch, and confirmed — this is a real hang, not just a logging weakness.

The enclosing loop is `while still_more:`, and both `still_more` and `docker_tag_url` are only assigned inside the `try`. If `raw_json["results"]` or `raw_json["next"]` raises, neither is updated, so the next pass re-fetches the identical URL, re-adds every tag on that page via `add_tag`, and logs again — forever.

Worth noting this pre-dates the branch: the handler was previously `log.debug`, so the loop span silently. Raising the log level to `warning` is what made it visible.

Fixed in `4aface4` by raising rather than falling through, consistent with the throttling path just above which already raises `requests.HTTPError`:

```python
except Exception as exc:
    # Leaving still_more and docker_tag_url unchanged here would
    # re-fetch the same page forever, re-adding tags each time.
    log.error("Malformed tag response for %s: %s", combined_repo_name, exc)
    raise requests.HTTPError(
        f"Malformed tag response from Docker Hub for {combined_repo_name}"
    ) from exc
```

I kept the catch surface as `Exception` rather than narrowing to `KeyError`/`TypeError`, so this commit changes only the control flow and not which failures are caught.

File: lftools_uv/api/endpoints/gerrit.py [hunk 2]
```diff
@@ -488,17 +489,17 @@ def create_project(
...
         try:
             results_dict = json.loads(json_text)  # pyright: ignore[reportAny]
-        except json.decoder.JSONDecodeError:
+        except json.decoder.JSONDecodeError as exc:
             log.info(resp)
-            log.info("A problem was encountered while querying the Gerrit API.")
+            log.error("A problem was encountered while querying the Gerrit API: %s", exc)
             log.debug(resp.text)
-            exit(resp.status_code)
+            sys.exit(resp.status_code)
```
[the remark this answers is not in the document]
Comment (ModeSevenIndustrialSolutions) [reply lftools_uv/api/endpoints/gerrit.py]: Agreed on the fix, with one correction to the reasoning.

An HTTP status can never be `0` coming from `requests`, so `sys.exit(resp.status_code)` would not actually report success — `sys.exit(200)` is non-zero and CI would see a failure. The real problems are different, and still worth fixing:

- Process exit codes are 0–255, so any status ≥ 256 would wrap. `sys.exit(200)` also collides with no convention anything downstream could interpret.
- As you say, the status is unrelated to the failure being reported. A body that fails to parse can arrive with **any** status, so echoing it tells the caller nothing about what went wrong.

Changed to `sys.exit(1)` in `4aface4`, with the reasoning recorded inline:

```python
# Not resp.status_code: an unparsable body can arrive with any
# status, and an HTTP code is not a meaningful process exit code.
sys.exit(1)
```

Note this line is pre-existing behaviour that this branch only touched to add the exception to the log, so it is a small scope increase — but it is squarely in the spirit of the branch, which is making failure paths say what actually happened.

File: lftools_uv/openstack/stack.py [hunk 3]
```diff
@@ -158,9 +158,12 @@ def _is_nested(resource: Any) -> bool:
...
     except Exception:
-        log.exception("Error calculating stack cost")
-        log.warning("Returning 0 total cost due to error")
-        print("total: 0.0")
+        # The per-server pricing lookup above already falls back to 0.0 on a
+        # flaky pricing API, so reaching here means stack enumeration itself
+        # failed and no total can be derived. Reporting 0.0 would be
+        # indistinguishable from a genuinely idle stack, so fail instead.
+        log.exception("Unable to calculate cost for stack %s", stack_name)
+        sys.exit(1)
```
[the remark this answers is not in the document]
Comment (ModeSevenIndustrialSolutions) [reply lftools_uv/openstack/stack.py:166]: Entirely fair — I called this behaviour change out in the PR description but did not cover it, which is exactly the gap the constitution's Principle I is meant to close. There was no `tests/test_openstack_stack.py` at all.

Added in `4aface4`, with both sides of the branch pinned:

```python
def test_cost_exits_non_zero_when_enumeration_fails(mock_connect, capsys, caplog):
    mock_cloud.orchestration.resources.side_effect = RuntimeError("enumeration failed")

    with pytest.raises(SystemExit) as exc_info:
        os_stack.cost("test-cloud", "broken-stack")

    assert exc_info.value.code == 1
    assert "total:" not in capsys.readouterr().out
    assert any(record.levelno >= logging.ERROR for record in caplog.records)
```

The `"total:" not in ...` assertion is the one that matters most: it pins the actual regression risk, which is not the exit code on its own but a fabricated total being printed alongside it.

A companion test covers the genuine empty case — a stack with no servers still prints `total: 0.0` and exits successfully — so the distinction between "nothing to bill" and "could not determine" stays enforced.

826 tests pass.

**Review by tykeal — APPROVED** [review]

**Comment by ModeSevenIndustrialSolutions** [comment]
## Diagnosis: the gate is blocking on pre-existing findings, not on anything this PR adds

Every check passes except `AI Slop Scan 🧹 / Audit changes`:

> `aislop quality gate failed: 38 finding(s) on the changed files (high 0, medium 38, low 0); a pull request must not add findings.`

### This PR adds nothing — it removes 25

Scanning the **same 19 files** with and without this branch:

| | findings |
|---|---:|
| `main` (without this PR) | **63** |
| this branch | **38** |

It removes 23 `silent-recovery`, 1 `python-broad-except` and 1 `print-debug`, and introduces none. The 38 that remain are pre-existing: `print()` calls in `openstack/stack.py` and `openstack/cmd.py`, and the hardcoded URLs in `stack.py` / `deploy.py` — all of which are fixed later in this same stack, by #647 and #648.

### Why the gate fails anyway

`AISLOP_GATE_LEVEL` is unset, so the org workflow defaults to **`any`**. That tier is not "don't add findings" — it is "every file you touch must be completely clean". The workflow's own comment says so:

> `any (default)` — Block the run when the pull request's changed files carry ANY finding, at any severity … a finding here is one this pull request adds, **or leaves behind in code it touched**.

`aislop --changes` scans the whole content of each changed file, not the diff hunks. So a PR inherits every finding in every file it opens.

### This is not specific to this stack

On `main`, **41 of 136 tracked Python files** carry at least one finding. Under `any`, a pull request touching any one of them is blocked until that entire file is clean — including a one-line bug fix. `lftools_uv/deploy.py` (7 findings, 992 lines) and `lftools_uv/openstack/stack.py` (24) are effectively frozen.

The `any` tier is a sound policy for a repository already at zero. This one has a backlog, which is what these PRs exist to clear.

### Gate status across the stack

| PR | findings on changed files | at `any` | at `high` |
|---|---:|---|---|
| #646 | 38 | ❌ | ✅ |
| #647 | 63 | ❌ | ✅ |
| #648 | 3 | ❌ | ✅ |
| #649 | 3 | ❌ | ✅ |
| #650 | 0 | ✅ | ✅ |
| #651 | 0 | ✅ | ✅ |
| #652 | 0 | ✅ | ✅ |

**Zero high-severity findings anywhere in the stack** — all 38/63/3/3 are medium.

The residuals on #648 and #649 are instructive: they are `helper_list` and `deploy.py`/`deploy_s3` complexity findings, cleared by #650's threshold config and by the `deploy.py` split that is still outstanding. So the ordering cannot be rearranged out of the problem — the last PRs to touch `deploy.py` will keep failing `any` until `deploy.py` is split, which needs characterisation tests for `deploy_s3` first (it currently has **none**).

### Recommendation

Set the repository variable **`AISLOP_GATE_LEVEL=high`** while the backlog is cleared, then **remove it** once the count reaches zero, returning the repo to the strict org default.

This is precisely what the gate's own error message instructs, and what the workflow documents:

> `high` — … The previous default; **use it as a per-repository escape hatch while a backlog is remediated.**
>
> A repository variable of the same name overrides the organisation value, so **a repository carrying a pre-existing backlog can sit at `high`** without weakening the estate-wide default.

It is safe here: there are no high-severity findings today, so `high` still blocks any newly introduced error-severity issue, and every PR in the stack is independently verified to reduce the count. Once the remaining six findings are gone, `any` becomes sustainable and should be restored — at that point it will do exactly what it is meant to.

### Alternative I considered and rejected

Reorganising the stack **by file** rather than **by rule**, so each PR leaves its files spotless. It would satisfy `any`, but the first PR touching `deploy.py` would have to also split the module — which needs tests for `deploy_s3` written first — and the first touching `openstack/` would absorb the whole print-to-logging change. That collapses seven reviewable PRs into two very large ones, exceeding the 20,000-line Copilot review limit this stack was shaped to stay under, and discards the review already done.

I have not changed any repository settings; that is your call.
